File: src/main/java/org/flighthub/Repository/FlightRepository.java
package org.flighthub.Repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.flighthub.Domain.Flight;
import org.flighthub.Utils.JdbcUtils;

import java.sql.*;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

public class FlightRepository implements IFlightRepository {
    private final JdbcUtils JDBCconnection;

    private static final Logger logger = LogManager.getLogger();

    public FlightRepository(JdbcUtils connection) {
        this.JDBCconnection = connection;
    }

    @Override
    public Optional<Flight> findOne(UUID id) {
        try {
            logger.traceEntry();
            logger.info("trying to find one Flight");
            Connection connection = JDBCconnection.getConnection();
            String query = "SELECT * FROM Flight WHERE id = ?";
            PreparedStatement statement = connection.prepareStatement(query);
            statement.setString(1, id.toString());
            ResultSet resultSet = statement.executeQuery();

            if (resultSet.next()) {
                Flight flight = new Flight(resultSet.getString("destination"),
                        resultSet.getObject("dateTime", LocalDateTime.class),
                        resultSet.getInt("availableSeats"),resultSet.getString("airport"));
                flight.setId(UUID.fromString(resultSet.getString("id")));
                return Optional.of(flight);
            }
        } catch (SQLException e) {
            System.out.println("Error Repo " + e);
            logger.error(e);
        }

        return Optional.empty();
    }
    @Override
    public Iterable<Flight> findByDestinationAndDepartureDate(String destination, LocalDate departureDate) {
        List<Flight> flights = new ArrayList<>();

        try {
            logger.traceEntry();
            logger.info("trying to find flights by destination and departure date");


            Connection connection = JDBCconnection.getConnection();
           // String query = "SELECT * FROM Flight WHERE destination = ? AND DATE(dateTime) = ?";
            String query="SELECT * FROM Flight WHERE destination = ? AND CAST(dateTime AS DATE) = ?";
            PreparedStatement statement = connection.prepareStatement(query);
            statement.setString(1, destination);
            statement.setDate(2, Date.valueOf(departureDate)); // Folosim doar data
            ResultSet resultSet = statement.executeQuery();

            while (resultSet.next()) {
                Flight flight = new Flight(resultSet.getString("destination"),
                        resultSet.getObject("dateTime", LocalDateTime.class),
                        resultSet.getInt("availableSeats"),resultSet.getString("airport"));
                flight.setId(UUID.fromString(resultSet.getString("id")));
                flights.add(flight);
            }
        } catch (SQLException e) {
            System.out.println("Error Repo " + e);
            logger.error(e);
        }

        return flights;
    }


    @Override
    public Iterable<Flight> findAll() {
        List<Flight> flights = new ArrayList<>();

        try {
            logger.traceEntry();
            logger.info("trying to find multiple Flights");
            Connection connection = JDBCconnection.getConnection();
            String query = "SELECT * FROM Flight";
            PreparedStatement statement = connection.prepareStatement(query);
            ResultSet resultSet = statement.executeQuery();

            while (resultSet.next()) {
                Flight flight = new Flight(resultSet.getString("destination"),
                        resultSet.getObject("dateTime", LocalDateTime.class),
                        resultSet.getInt("availableSeats"),resultSet.getString("airport"));
                flight.setId(UUID.fromString(resultSet.getString("id")));
                flights.add(flight);
            }
        } catch (SQLException e) {
            System.out.println("Error Repo " + e);
            logger.error(e);
        }

        return flights;
    }

    @Override
    public Flight save(Flight entity) {
        try {
            logger.traceEntry();
            logger.info("trying to save one Flight");
            Connection connection = JDBCconnection.getConnection();
            String query = "INSERT INTO Flight (id, destination, dateTime, availableSeats, airport) VALUES (?, ?,?, ?, ?)";
            PreparedStatement statement = connection.prepareStatement(query);
            UUID uuid = UUID.randomUUID();
            entity.setId(uuid);
            statement.setString(1, uuid.toString());
            statement.setString(2, entity.getDestination());
            statement.setObject(3, entity.getDateTime());
            statement.setInt(4, entity.getAvailableSeats());
            statement.setString(5,entity.getAirport());
            statement.executeUpdate();
        } catch (SQLException e) {
            System.out.println("Error Repo " + e);
            logger.error(e);
        }

        return entity;
    }

    @Override
    public Optional<Flight> delete(UUID id) {
        Optional<Flight> flightOptional = findOne(id);

        if (flightOptional.isPresent()) {
            try {
                logger.traceEntry();
                logger.info("trying to delete one Flight");
                Connection connection = JDBCconnection.getConnection();
                String query = "DELETE FROM Flight WHERE id = ?";
                PreparedStatement statement = connection.prepareStatement(query);
                statement.setString(1, id.toString());
                statement.executeUpdate();
            } catch (SQLException e) {
                System.out.println("Error Repo " + e);
                logger.error(e);
            }
        }

        return flightOptional;
    }

    @Override
    public Optional<Flight> update(Flight entity) {
        Optional<Flight> flightOptional = findOne(entity.getId());
        /*
        if (flightOptional.isPresent()) {
            try {
                logger.traceEntry();
                logger.info("trying to update one Flight");
                Connection connection = JDBCconnection.getConnection();
                String query = "UPDATE Flight SET destination = ?, dateTime = ?, availableSeats = ? WHERE id = ?";
                PreparedStatement statement = connection.prepareStatement(query);
                statement.setString(1, entity.getDestination());
                statement.setObject(2, entity.getDateTime());
                statement.setInt(3, entity.getAvailableSeats());
                statement.setString(4, entity.getId().toString());
                statement.executeUpdate();
            } catch (SQLException e) {
                System.out.println("Error Repo " + e);
                logger.error(e);
            }
        }
*/
        return flightOptional;
    }
    @Override
    public void updateAvailableSeats(UUID flightId, int newAvailableSeats) {
        try {
            logger.traceEntry();
            logger.info("trying to update available seats for Flight");
            Connection connection = JDBCconnection.getConnection();
            String query = "UPDATE Flight SET availableSeats = ? WHERE id = ?";
            PreparedStatement statement = connection.prepareStatement(query);
            statement.setInt(1, newAvailableSeats);
            statement.setString(2, flightId.toString());
            statement.executeUpdate();
        } catch (SQLException e) {
            System.out.println("Error Repo " + e);
            logger.error(e);
        }
    }
}

File: Server - C#/persistence/FlightRepository.cs
using FlightHubC_.Domain;
using FlightHubC_.Utils;
//using log4net;
using System;
using System.Collections.Generic;
using System.Data.SqlClient;

namespace FlightHubC_.Repository
{
    public class FlightRepository : IFlightRepository
    {
        private readonly SqlUtils sqlUtils;
        //private static readonly log4net.ILog log = LogManager.GetLogger(typeof(Program));

        public FlightRepository(SqlUtils sqlUtils)
        {
            this.sqlUtils = sqlUtils;
        }

        public Flight FindOne(Guid id)
        {
            //log.Info("finding one Flight");
            Flight flight = null;
            string query = "SELECT * FROM Flight WHERE id = @Id";

            using (SqlConnection connection = sqlUtils.GetConnection())
            {
                using (SqlCommand command = new SqlCommand(query, connection))
                {
                    command.Parameters.AddWithValue("@Id", id);
                    connection.Open();
                    SqlDataReader reader = command.ExecuteReader();
                    if (reader.Read())
                    {
                        flight = new Flight(
                            reader["destination"].ToString(),
                            Convert.ToDateTime(reader["dateTime"]),
                            Convert.ToInt32(reader["availableSeats"]),
                            reader["airport"].ToString()
                        );
                        flight.Id = id;
                    }
                }
            }

            return flight;
        }

        public IEnumerable<Flight> FindByDestinationAndDepartureDate(string destination, DateTime departureDate)
        {
            //log.Info("searching flight by destination and departure date");

            List<Flight> flights = new List<Flight>();
            string query = "SELECT * FROM Flight WHERE destination = @Destination AND CAST(dateTime AS DATE) = @DepartureDate AND availableSeats > 0";

            using (SqlConnection connection = sqlUtils.GetConnection())
            {
                using (SqlCommand command = new SqlCommand(query, connection))
                {
                    command.Parameters.AddWithValue("@Destination", destination);
                    command.Parameters.AddWithValue("@DepartureDate", departureDate.Date);
                    connection.Open();
                    SqlDataReader reader = command.ExecuteReader();
                    while (reader.Read())
                    {
                        Flight flight = new Flight(
                            reader["destination"].ToString(),
                            Convert.ToDateTime(reader["dateTime"]),
                            Convert.ToInt32(reader["availableSeats"]),
                            reader["airport"].ToString()
                        );
                        flight.Id = reader.GetGuid(reader.GetOrdinal("id"));
                        flights.Add(flight);
                    }
                }
            }

            return flights;
        }
        public Flight FindByDetails(string destination, DateTime departureDateTime, int availableSeats, string airport)
        {
            //log.Info("searching flight by details");
            Flight foundFlight = null;
            string query = "SELECT * FROM Flight WHERE destination = @Destination " +
                           "AND CAST(dateTime AS DATE) = @DepartureDateTime " +
                           "AND availableSeats = @AvailableSeats " +
                           "AND airport = @Airport";

            using (SqlConnection connection = sqlUtils.GetConnection())
            {
                using (SqlCommand command = new SqlCommand(query, connection))
                {
                    command.Parameters.AddWithValue("@Destination", destination);
                    command.Parameters.AddWithValue("@DepartureDateTime", departureDateTime.Date);
                    command.Parameters.AddWithValue("@AvailableSeats", availableSeats);
                    command.Parameters.AddWithValue("@Airport", airport);
                    connection.Open();
                    SqlDataReader reader = command.ExecuteReader();
                    if (reader.Read())
                    {
                        foundFlight = new Flight(
                            reader["destination"].ToString(),
                            Convert.ToDateTime(reader["dateTime"]),
                            Convert.ToInt32(reader["availableSeats"]),
                            reader["airport"].ToString()
                        );
                        foundFlight.Id = reader.GetGuid(reader.GetOrdinal("id"));
                    }
                }
            }

            return foundFlight;
        }


        public IEnumerable<Flight> FindAll()
        {

            List<Flight> flights = new List<Flight>();
            string query = "SELECT * FROM Flight WHERE availableSeats > 0";

            using (SqlConnection connection = sqlUtils.GetConnection())
            {
                using (SqlCommand command = new SqlCommand(query, connection))
                {
                    //log.Info("finding all Flight");
                    connection.Open();
                    SqlDataReader reader = command.ExecuteReader();
                    while (reader.Read())
                    {
                        Flight flight = new Flight(
                        reader["destination"].ToString(),
                        Convert.ToDateTime(reader["dateTime"]),
                        Convert.ToInt32(reader["availableSeats"]),
                        reader["airport"].ToString()
                    );
                        flight.Id = reader.GetGuid(reader.GetOrdinal("id"));
                        flights.Add(flight);
                    }
                }
            }

            return flights;
        }

        public Flight Save(Flight entity)
        {
            string query = "INSERT INTO Flight (id, destination, dateTime, availableSeats, airport) " +
                           "VALUES (@Id, @Destination, @DateTime, @AvailableSeats, @Airport)";

            using (SqlConnection connection = sqlUtils.GetConnection())
            {
                using (SqlCommand command = new SqlCommand(query, connection))
                {
                    //log.Info("saving one Flight");
                    Guid guid = Guid.NewGuid();
                    entity.Id = guid;
                    command.Parameters.AddWithValue("@Id", entity.Id);
                    command.Parameters.AddWithValue("@Destination", entity.Destination);
                    command.Parameters.AddWithValue("@DateTime", entity.DateTime);
                    command.Parameters.AddWithValue("@AvailableSeats", entity.AvailableSeats);
                    command.Parameters.AddWithValue("@Airport", entity.Airport);
                    connection.Open();
                    command.ExecuteNonQuery();
                }
            }

            return entity;
        }

        public Flight Delete(Guid id)
        {
            Flight deletedFlight = FindOne(id);
            if (deletedFlight != null)
            {
                string query = "DELETE FROM Flight WHERE id = @Id";

                using (SqlConnection connection = sqlUtils.GetConnection())
                {
                    using (SqlCommand command = new SqlCommand(query, connection))
                    {
                        command.Parameters.AddWithValue("@Id", id);
                        connection.Open();
                        command.ExecuteNonQuery();
                    }
                }
            }

            return deletedFlight;
        }

        public Flight Update(Flight entity)
        {


            return entity;
        }
        public void updateAvailableSeats(Guid flightId, int newAvailableSeats)
        {
            //log.Info("updating available seats for one Flight");
            String query = "UPDATE Flight SET availableSeats = @New WHERE id = @Id";
            using (SqlConnection connection = sqlUtils.GetConnection())
            {
                using (SqlCommand command = new SqlCommand(query, connection))
                {
                    command.Parameters.AddWithValue("@New", newAvailableSeats);
                    command.Parameters.AddWithValue("@Id", flightId);
                    connection.Open();
                    command.ExecuteNonQuery();
                }
            }

        }
    }
}
